File: server/datasets/python_source-code/Multiple-E2/type2_string_clone_1_valid_parentheses.py
def validate_sequence(text):
    memory_string = ""
    matches = {")": "(", "}": "{", "]": "["}
    
    for letter in text:
        if letter not in matches:
            memory_string += letter
        else:
            if len(memory_string) == 0:
                return False
            if memory_string[-1] != matches[letter]:
                return False
            memory_string = memory_string[:-1]
            
    return len(memory_string) == 0

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_e2/type2_string_clone_1_valid_parentheses.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public boolean validate_sequence(String text) {
        StringBuilder memory_string = new StringBuilder();
        Map<Character, Character> matches = new HashMap<>();
        matches.put(')', '(');
        matches.put('}', '{');
        matches.put(']', '[');
        
        for (char letter : text.toCharArray()) {
            if (!matches.containsKey(letter)) {
                memory_string.append(letter);
            } else {
                if (memory_string.length() == 0) {
                    return false;
                }
                if (memory_string.charAt(memory_string.length() - 1) != matches.get(letter)) {
                    return false;
                }
                memory_string.deleteCharAt(memory_string.length() - 1);
            }
        }
        return memory_string.length() == 0;
    }
}
